### PR TITLE
Agent inspection on mouse hover

### PR DESCRIPTION
**Pull request by fbanning**
Closes https://github.com/JuliaDynamics/Agents.jl/issues/538

This turned out to be a bigger endeavour than initially anticipated. I had to refactor the codebase a bit and most notably had to introduce a new Makie plot recipe.

The recipe introduces the `ABMPlot` type and its new generated functions `abmplot` and `abmplot!` which takes a list of agent positions as well as the model object and converts them into observables. Those are then accessible via the plot object itself (e.g. `plot.model[]`). The new type can be extended at will and we can easily create new methods for different use cases, e.g. for `GraphSpace` models later on.

## Showcase

Same start for every showcase example: 
```julia
using Agents
using InteractiveDynamics
using GLMakie

model, agent_step!, model_step! = Models.schelling()
```

That means I've only tested a simple 2D GridSpace model for now.

Then we try out the three main plotting functions.

```julia
## abm_plot with kwargs

ac(a) = a.group == 1 ? :red : :blue
fig, abmstepper, inspector = abm_plot(model; ac=ac)
fig
```
abm_plot

```julia
## abm_play

fig, abmstepper, inspector = abm_play(model, agent_step!, model_step!)
fig
```
https://user-images.githubusercontent.com/65158285/129541861-1f39a78a-26d2-4023-b387-55a8eafd1c70.mp4

```julia
## abm_data_exploration
using Statistics

fig, adf, mdf = abm_data_exploration(model, agent_step!, model_step!, Dict(); 
    adata=[(:mood, mean)])
```
https://user-images.githubusercontent.com/65158285/129541896-e1f84100-f9e0-44df-b0a1-647de57c8121.mp4

Here we can see why it was necessary to introduce the new `ABMPlot` type. We can inspect scatter points as agents on the top left axis but still regularly inspect the scatter points on the observed variables on the right hand side. :)

## Sidenote

I've also introduced the possibility to use a different colourscheme than the default one. Sorry about mixing it into this PR. Realised it a bit too late that it would have been nicer to split it into another PR. It's not a very big thing but is something that was on my to-do list for a while. :)

## To do
- [x] 3D models
    - [x] Implement 3D plotting method (untested basics are already there)
    - [x] Test
- [x] 2D model with polygons
    - [x] Implement poly plotting method
    - [x] Test 
    - [x] Disable inspector until properly implemented
- [x] Test with a `ContinuousSpace` model
- [x] Test `abm_video` for 2D, 3D and poly plotting
- [x] Truncate most floats
- [x] Allow custom agent strings for tooltips.
- [x] Update docstrings and add documentation for new features
- [x] Test and document colorscheme keyword

## Further ideas
- As proposed [here](https://github.com/JuliaDynamics/Agents.jl/issues/538#issuecomment-896059948), we could improve the tooltip formatting in cases with multiple agents on one position.
- Maybe `modelobs` is unnecessary now that we could just use `plot.model[]`?
- Maybe some of the `Observable` conversions (`ac`, `am`, `as`, etc.) can be dropped when they're integrated into the plot object as Attributes. Not sure about this as it might mess with the `abmstepper` flow.

That's it from my side. Happy to answer any questions, discuss function naming as well as your ideas for extensions/changes. :)

### CODE REVIEW

**Comment by fbanning** [comment]
I probably won't have time to work on this until the end of August. I've enabled the ability for repo maintainers to add commits, so please feel free to enhance, change and/or merge at will. :)

**Comment by Datseris** [comment]
I'm a bit behind here. So if I merge this, the polygon plotting goes away?

**Comment by fbanning** [comment]
No, polygon plots haven't been removed. They are just not yet

1. integrated into the new `ABMPlot` recipe system, meaning that there's no specialised inspector functionality for the `poly!` plots
2. tested at all

I would advise against merging this PR right now until the To-Do list above has been worked on. I will mark the PR as Draft for now, just in case.

**Comment by fbanning** [comment]
I've just tested a poly model, namely the one in `Agents.Models.flocking`. Plotting is working just fine although the mouse hover tooltip is of course not very helpful and just displays the coordinates of the bounding box of all polys.

`abm_plot`:

abm_plot

`abm_play`:

https://user-images.githubusercontent.com/65158285/132388269-d3cf0e7c-5397-42fc-a13e-899a3585a6d8.mp4

`abm_data_exploration`:

https://user-images.githubusercontent.com/65158285/132388416-58d3b48f-2515-42aa-86e6-3030e19e130d.mp4

And here's a screenshot of the not-so-helpful-yet tooltip on mouse hover:

grafik

So yes, I can confirm that poly plots still work as they did before, although they aren't integrated into the new `ABMPlot` recipe system yet (which results in these useless tooltips).

But it's definitely not a showstopper and shouldn't block the PR from being merged.

However, if someone wants to contribute the recipe for poly plots, feel free to do so. :)

**Comment by fbanning** [comment]
I've added some handling for `ContinuousSpace` models and tested it with the integrated `Agents.Models.social_distancing` model.

`abm_plot`:

abm_plot

`abm_play`:

https://user-images.githubusercontent.com/65158285/132408487-e94ef16c-3a27-4e88-9024-ebd7ba2e3dc9.mp4

`abm_data_exploration`:

https://user-images.githubusercontent.com/65158285/132408508-34bf76ea-0a4d-438f-988e-bec38b128395.mp4

Everything seems to be in order from what I can tell.

**Comment by fbanning** [comment]
`abm_video` should work as well.

Schelling:

https://user-images.githubusercontent.com/65158285/132415631-cc956f10-885b-44e7-9447-8856e46a3e28.mp4

Social Distancing:

https://user-images.githubusercontent.com/65158285/132415666-82bbf5e2-6110-4d42-9337-442cdf8de3f2.mp4

Flocking:

https://user-images.githubusercontent.com/65158285/132415998-c1330b47-f16e-46d3-9f01-2babd9267c80.mp4

Checking this off the list. ✅

**Comment by Datseris** [comment]
I think for continjous space the mouseover should round position to 3 (or even 2) sigdigits via `round`

**Comment by fbanning** [comment]
3D plots are now implemented and tested as well. I've used the [Rabbit, Fox, Hawk](https://juliadynamics.github.io/Agents.jl/stable/examples/rabbit_fox_hawk/) example from our docs.

`abm_plot`:
abm_plot

`abm_play`:

https://user-images.githubusercontent.com/65158285/132552887-aeabc19d-85a8-489d-bfd9-d13021cb3ad4.mp4

`abm_data_exploration`:

https://user-images.githubusercontent.com/65158285/132552894-65bb4304-5200-4c00-a33a-7f1b4a677131.mp4

`abm_video`:

https://user-images.githubusercontent.com/65158285/132552912-463e0d9e-c3ea-455f-ae1c-7a7f1cd12611.mp4

**Comment by fbanning** [comment]
I've also disabled the inspector for static preplots because those are just "backdrops" and shouldn't be inspected at all. Inspection only makes sense for the actual data, not the terrain/surroundings/whatever.

**Comment by fbanning** [comment]
(Sorry for the ping if inappropriate.)  
@Datseris or @Libbum: From my perspective, this PR is now ready for an in-depth review, if you want to, and possibly even a merge in the end. :)

**Comment by Datseris** [comment]
pings are never innapropriate. My only problem is that at the moment I cannot do a full in depth review due to time constraints in other projects :( If @Libbum can do it that's great. Nevertheless, the plots you show are all satisfactory for me and I'm going to merge in a couple of days if Tim haven't gotten around for a deeper review.

**Comment by Libbum** [comment]
I'm taking next week off of work to decompress and ultimately catch up on things like this. Will take a look then if that's OK.

**Comment by Libbum** [comment]
@fbanning, I can see a number of #TODO tags in the comments, most are probably done now. Can you clean those up so it's just the ones (if any) that should remain is all that's in the code?

**Comment by fbanning** [comment]
>  Can you clean those up so it's just the ones (if any) that should remain is all that's in the code?

Done.

>  Will take a look then if that's OK.

Awesome, thanks for taking the time but remember to not fill your days off with other work again. Relaxation should be your top priority during that week. :)

**Comment by fbanning** [comment]
I've also integrated the polygon plots into the `ABMPlot` recipe system. The plotted output remains the same as before (see images/videos in comment above). Note that the tooltip for polygon plots remains unhelpful - something to look into in the future. This now ticks off the last point from the todo list above. 🚀

**Comment by Datseris** [comment]
@Libbum should I merge this?

**Comment by Datseris** [comment]
I confirm this works with DaisyWorld: 

image

If I hover over a Diasy I get the correct stuff. If I hover over the heatmap, I get the heatmap value. Very nice! Although, I'm not sure why it is named `H`. @fbanning where is this name decided?

**Comment by Libbum** [comment]
Some issues for flocking:
1. We may need to mediate how many properties are shown
2. We probably need to truncate floats as well
image
3. Is there a way we can fit the tooltip within the global window?
image
4. Failures with custom polygons
image
```julia
Error in callback:
type PixelCamera has no field eyeposition
Stacktrace:
  [1] getproperty(x::Makie.PixelCamera, f::Symbol)
    @ Base ./Base.jl:33
  [2] show_data(inspector::DataInspector, plot::Mesh{Tuple{GeometryBasics.Mesh{2, Float32, GeometryBasics.Ngon{2, Float32, 3, Point{2, Float32}}, GeometryBasics.SimpleFaceView{2, Float32, 3, GeometryBasics.OffsetInteger{-1, UInt32}, Point{2, Float32}, GeometryBasics.NgonFace{3, GeometryBasics.OffsetInteger{-1, UInt32}}}}}}, idx::UInt32)
    @ Makie ~/.julia/packages/Makie/ZK3qM/src/interaction/inspector.jl:627
  [3] show_data_recursion(inspector::DataInspector, plot::Mesh{Tuple{GeometryBasics.Mesh{2, Float32, GeometryBasics.Ngon{2, Float32, 3, Point{2, Float32}}, GeometryBasics.SimpleFaceView{2, Float32, 3, GeometryBasics.OffsetInteger{-1, UInt32}, Point{2, Float32}, GeometryBasics.NgonFace{3, GeometryBasics.OffsetInteger{-1, UInt32}}}}}}, idx::UInt32)
    @ Makie ~/.julia/packages/Makie/ZK3qM/src/interaction/inspector.jl:482
  [4] on_hover(inspector::DataInspector)
    @ Makie ~/.julia/packages/Makie/ZK3qM/src/interaction/inspector.jl:458
  [5] (::Makie.var"#869#871"{DataInspector})(#unused#::Tuple{Float64, Float64})
    @ Makie ~/.julia/packages/Makie/ZK3qM/src/interaction/inspector.jl:443
  [6] #invokelatest#2
    @ ./essentials.jl:708 [inlined]
  [7] invokelatest
    @ ./essentials.jl:706 [inlined]
  [8] notify(observable::Makie.PriorityObservable{Tuple{Float64, Float64}})
    @ Makie ~/.julia/packages/Makie/ZK3qM/src/interaction/PriorityObservable.jl:72
  [9] setindex!
    @ ~/.julia/packages/Makie/ZK3qM/src/interaction/PriorityObservable.jl:66 [inlined]
 [10] (::GLMakie.var"#71#72"{Events, GLFW.Window})(#unused#::Nothing)
    @ GLMakie ~/.julia/packages/GLMakie/ayWSm/src/events.jl:6
 [11] #invokelatest#2
    @ ./essentials.jl:708 [inlined]
 [12] invokelatest
    @ ./essentials.jl:706 [inlined]
 [13] notify
    @ ~/.julia/packages/Observables/OFj0u/src/Observables.jl:88 [inlined]
 [14] setindex!(observable::Observable{Nothing}, val::Nothing)
    @ Observables ~/.julia/packages/Observables/OFj0u/src/Observables.jl:248
 [15] fps_renderloop(screen::GLMakie.Screen, framerate::Float64)
    @ GLMakie ~/.julia/packages/GLMakie/ayWSm/src/rendering.jl:22
 [16] renderloop(screen::GLMakie.Screen; framerate::Float64)
    @ GLMakie ~/.julia/packages/GLMakie/ayWSm/src/rendering.jl:48
 [17] renderloop(screen::GLMakie.Screen)
    @ GLMakie ~/.julia/packages/GLMakie/ayWSm/src/rendering.jl:41
 [18] (::GLMakie.var"#50#52"{GLMakie.Screen})()
    @ GLMakie ./task.jl:406
```

**Comment by Datseris** [comment]
Custom polygons are tricky, and hopefully I'd like to merge this in before Monday for showcase in a tutorial. Can we just add a simple clause that does not enable tooltips for polygon plotting?

**Comment by Libbum** [comment]
Ah, sorry, yes - I hadn't read through all of the thread completely. Sure, we can try dropping that for the moment.

**Comment by Datseris** [comment]
Actually it seems there are still some problems. I just run this in the Schelling model where I change each agent's plot style and I see wrong clipping:

image


with
```
groupcolor(agent) = agent.group == 1 ? :blue : :orange
groupmarker(agent) = agent.group == 1 ? :circle : :rect

fig, _ = abm_plot(model; ac = groupcolor, am = groupmarker)
display(fig)
```

**Review by Libbum — APPROVED** [review]
In principle, things are looking like they work as a first pass. I've checked multiple grid and continuous space examples and they all handle pretty well. I'd be happy to merge essentially early to have this available for the tutorial.

- **Before merge**:  @fbanning, could you turn off polygon tooltips? I tried removing 
```julia
function Makie.show_data(inspector::DataInspector,
            plot::ABMPlot{<:Tuple{Vector{<:Polygon{2}}, <:Agents.ABM}},
            idx, ::Makie.Poly)
```
But still got the failures I had before with flocking.

For the next iteration, some comments:
- The `H` for heatmap - no idea where this is coming from, but it'd be good to track it down and have a keyword to name it. 
image
Could be `Elevation[135, 339]` for example
- We need better parsing for more complicated data. I wonder if we can actually allow the user to provide a function like `show` of their agent?
- Aligning the tooltip to sit within the global window would be great.

**Comment by Libbum** [comment]
@Datseris, Oh! Yeah that's problematic.

**Comment by fbanning** [comment]
> I'm not sure why it is named `H`. @fbanning where is this name decided?

For the heatmap tooltip, we're using the built-in method from Makie itself. That means the `H[i,j] = val` [name comes from there](https://github.com/JuliaPlots/Makie.jl/blob/6b71a7f0b60aa0573309a4f236b9d5c56c4d4cc8/src/interaction/inspector.jl#L722) and the used `show_imagelike` function. If you have a nice idea how the users can provide an alternative String (like "Elevation" or whatever), then this shouldn't be hard to implement, I suppose.

**Comment by fbanning** [comment]
>     * could you turn off polygon tooltips? I tried removing
> 
> 
> ```julia
> function Makie.show_data(inspector::DataInspector,
>             plot::ABMPlot{<:Tuple{Vector{<:Polygon{2}}, <:Agents.ABM}},
>             idx, ::Makie.Poly)
> ```
> 
> But still got the failures I had before with flocking.

I guess it should be possible to disable the inspector for now if the user provided polygons as markers. I've not done this because I thought maybe someone would be interested in writing a method for the poly plots in the near future (I personally don't use it and have no experience with poly plots at all). The process to disable the inspector just for the poly plots while leaving it enabled for the rest of the plots (e.g. scatterlines used in `abm_data_exploration`) should be the same as I did for the `static_preplot`. Need to check which point in the pipeline is the best to turn it off though.

**Comment by fbanning** [comment]
> * We need better parsing for more complicated data. I wonder if we can actually allow the user to provide a function like `show` of their agent?
> 
> * Aligning the tooltip to sit within the global window would be great.

Both of these are something that are on my list as well but haven't included them yet because this was already pretty time consuming as is. 😅

**Comment by fbanning** [comment]
> I just run this in the Schelling model where I change each agent's plot style and I see wrong clipping

Hm, this is weird and hasn't happened to me during my tests. Will try to reproduce it again.

Anyways, I think this looks like a problem with the `depth` value of the inspector (see [here](https://makie.juliaplots.org/v0.15.1/documentation/inspector/#attributes_of_datainspector)). By default it is set to `9e3` but perhaps we need to put it...

over 9000

**Comment by Libbum** [comment]
`H`: Ah, I see! Not sure we need to go that deep. It was confusing me that it was an 'obvious' choice, but could not be found in our code.

polygon tooltips and other features: No need to rush these anymore. We wanted to merge this on the weekend for the tutorial, but since we didn't end up doing that the time pressure is not so bad.

depth: I tried that locally, but setting it to 1e20 didn't really work at all. The plot even failed to show... Maybe I did it wrong though..

**Comment by fbanning** [comment]
> depth: I tried that locally, but setting it to 1e20 didn't really work at all. The plot even failed to show... Maybe I did it wrong though..

Yep, same result here. Once I use a function as input for `am`, I get this overlap problem. Simply providing a different marker (e.g. `am=:rect`) seems to work fine. Not sure yet what's the root of the problem here but I'll do some more digging.

**Comment by fbanning** [comment]
>     * We need better parsing for more complicated data. I wonder if we can actually allow the user to provide a function like `show` of their agent?

Just tested this and it is already easily doable. Users only need to define a specialised method of the `InteractiveDynamics.agent2string` function and this will already work.

```julia
InteractiveDynamics.agent2string(agent::Models.SocietyMember) = "This is an agent string."
```

grafik

**Comment by Libbum** [comment]
> define a specialised method of the InteractiveDynamics.agent2string function

Awesome! Lets export that function then, and provide a more 'complex' example to simplify the display of an agent like `Bird`.

This actually negates a lot of the floating point rounding requirements I was eluding to as well, since users can format their agent details as they like.

I assume this string can be mulitline right?

**Comment by fbanning** [comment]
> Awesome! Lets export that function then, and provide a more 'complex' example to simplify the display of an agent like `Bird`.

Done. Went with an imaginary example that should still display what to do without relying on a specific example.
 
> I assume this string can be mulitline right?

Sure thing.

**Comment by fbanning** [comment]
In b89cdb7 I've attempted to separate the initialisation of the abmstepper and the plots themselves. I think this makes the code a lot more maintainable in the long run and also contributes to keeping the functions a bit more "clean" as they now only have one major goal instead of doing multiple things at once. I've ran some tests (although not as extensive as all of those above) and everything seems to be in working order.

**Comment by fbanning** [comment]
Oh and in df542e2 I've changed the return values of the plotting functions back to what they were before. Initially I returned the `inspector` because I thought it might be helpful to have easy access to it but that's not really needed since it (the inspector) can be easily retrieved from the `fig` object. This avoids breaking other people's existing code.

**Comment by Libbum** [comment]
Thanks Fredrick! I'll have another once over tomorrow hopefully, but each of your changes sounds like a good idea.

**Comment by Datseris** [comment]
@fbanning have you contacted Makie people about this? In any case, to my understanding, the current functionality does not break anything, works for standard plotting and semi-works for function-as-marker plotting. It doesn't break it, it just shows the tooltip below the markers. 

So, I think we should be able to merge and tag this, and let the remaining as a bugfix issue?

**Comment by fbanning** [comment]
> have you contacted Makie people about this?

No, sorry, didn't find the time to do that. Will try to do it next week. Won't matter though if I refer to this PR or to another newly created bugfix issue.

> I think we should be able to merge and tag this

Yes, thanks, I think so too. However, documentation needs to be updated before merging this PR, otherwise users won't know that (a) this functionality is there and (b) they can easily change the tooltip with an own method for their agent types. If somebody wants to write this up, please feel free to do so. Otherwise I'll also try to find an hour for that next week.

**Comment by fbanning** [comment]
And I've just realised that I've never shown that the newly implemented `colorscheme` keyword actually works.

```julia
# same tests as above just with two additions
colorscheme = [:red, :blue, :green, :yellow, :grey]
ac(a) = a.group == 1 ? colorscheme[1] : colorscheme[2]

fig, adf, mdf = abm_data_exploration(model, agent_step!, model_step!, params; 
    adata=[(:mood, mean), (:mood, minimum)], ac, colorscheme)
```
with_colorscheme

I've already previously added it to the list of figure related keywords in the docstring of `abm_plot`, so it's documented enough, I think.

**Comment by fbanning** [comment]
@Datseris I think you can merge this now.

**Pull request merged**
